Objects/floatobject.c

```diff
@@ -1683,8 +1685,69 @@ typedef enum _py_float_format_type float_format_type;
 #define ieee_big_endian_format _py_float_format_ieee_big_endian
 #define ieee_little_endian_format _py_float_format_ieee_little_endian
 
-#define float_format (_PyRuntime.float_state.float_format)
-#define double_format (_PyRuntime.float_state.double_format)
+/* We attempt to determine if this machine is using IEEE
+    floating point formats by peering at the bits of some
+    carefully chosen values.  If it looks like we are on an
+    IEEE platform, the float packing/unpacking routines can
+    just copy bits, if not they resort to arithmetic & shifts
+    and masks.  The shifts & masks approach works on all finite
+    values, but what happens to infinities, NaNs and signed
+    zeroes on packing is an accident, and attempting to unpack
+    a NaN or an infinity will raise an exception.
+
+    Note that if we're on some whacked-out platform which uses
+    IEEE formats but isn't strictly little-endian or big-
+    endian, we will fall back to the portable shifts & masks
+    method.
+
+    These functions are reduced to a no-op via compiler
+    optimizations. */
+
+#ifdef __GNUC__
+__attribute__((always_inline))
+#endif
+static inline float_format_type
+get_float_format(void)
+{
+    if (sizeof(float) == 4) {
+        float y = 16711938.0;
+        uint32_t z;
+        memcpy(&z, &y, 4);
+        if (z == 0x4b7f0102)
+        {
+            if (memcmp(&y, "\x4b\x7f\x01\x02", 4) == 0) {
+                return ieee_big_endian_format;
+            }
+            else if (memcmp(&y, "\x02\x01\x7f\x4b", 4) == 0) {
+                return ieee_little_endian_format;
+            }
+        }
+    }
+    return unknown_format;
+}
+
+#ifdef __GNUC__
+__attribute__((always_inline))
+#endif
+static inline float_format_type
+get_double_format(void)
+{
+    if (sizeof(double) == 8) {
```

Comment:
Can't you use the `#if SIZEOF_DOUBLE == 8` instead as well? it would make the runtime faster (unless GCC makes some magical things here, but I don't know how the assembly looks like here).